### DISCUSSION TITLE
[Kafka] Is it possible to validate the content of a Kafka message containing an OpenAPI-generated object

### DISCUSSION THREAD

**Discussion by mfnztdkt** [Decathlon/tzatziki#398] [Q&A]
I tried to validate the content of a Kafka message the following way :
```
And the ... topic contains this ...:
    """yml
        groupOfFields1:
          - field1: value1
            field2: value2
          - field1: value3
            field2: value4
        groupOfFields2:
          - field3: value5
            field4: value6
          - field3: value7
            field4: value8
    """
```
It does not work for messages containing objects whose toString method does not use dotNotation (for example objects generated by OpenAPI).

```
com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `java.util.LinkedHashMap` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('class MyClass { [...]
 at [Source: (StringReader); line: 1, column: 1]
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
[...]
	at com.decathlon.tzatziki.utils.JacksonMapper.read(JacksonMapper.java:51)
	at com.decathlon.tzatziki.utils.Mapper.read(Mapper.java:27)
	at com.decathlon.tzatziki.steps.KafkaSteps.lambda$the_topic_contains$20(KafkaSteps.java:313)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
```

Do you know if there is a particular way to make it work ?

**Comment by brian-mulier-p** [comment]
Hello ! [In there](https://github.com/Decathlon/tzatziki/tree/main/tzatziki-spring-kafka#asserting-messages) you'll find the answer but basically everything but `json message` is considered as an avro-schema-based assertion. So to assert in your case you can use
```
And the ... topic contains this json message:
    """yml
        groupOfFields1:
          - field1: value1
            field2: value2
          - field1: value3
            field2: value4
        groupOfFields2:
          - field3: value5
            field4: value6
          - field3: value7
            field4: value8
    """
```

**Reply by mfnztdkt** [reply]
Sorry and thank you Brian for your quick answer, I continued the investigation and updated the description. It seems related to the non-standard toString implementation of an OpenAPI-generated object.

**Reply by mfnztdkt** [reply]
I see that Tzatziki tries to transform a Kafka record value into a map of values by passing record.value().toString() to a Mapper, and that mapper expects either a Json or Yaml or DotNotation representation. Do you have an idea on how to deal with any "non standard" toString implementation of Kafka record values ? I managed to write a customized KafkaSteps to make it work without using toString() method but I feel like it may not be the best solution.
